### PR TITLE
TreeView: fix selectByClick selection after changing showCheckboxesMode to 'selectAll' at runtime (T1190412)

### PR DESCRIPTION
**Pull request by AlexanderMoiseev**

### CODE REVIEW

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 1]
```diff
@@ -1432,6 +1442,22 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
+    },
+
+    _getEventsData: function() {
+        const clickSelector = '.' + this._itemClass();
+        const pointerDownSelector = '.' + NODE_CLASS + ', .' + SELECT_ALL_ITEM_CLASS;
+        const eventName = addNamespace(clickEventName, this.NAME);
+        const pointerDownEvent = addNamespace(pointerEvents.down, this.NAME);
+
+        return { eventName, clickSelector, pointerDownEvent, pointerDownSelector };
+    },
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js:1461]: [boy scout rule]
Let's refactor this old code as we change it:
```javascript
        const itemSelector = `.${this._itemClass()}`;
        const nodeSelector = `.${NODE_CLASS}, .${SELECT_ALL_ITEM_CLASS}`;
        const clickEventNamespace = addNamespace(clickEventName, this.NAME);
        const pointerDownEventNamespace = addNamespace(pointerEvents.down, this.NAME);

        return {
            clickEventNamespace,
            itemSelector,
            pointerDownEventNamespace,
            nodeSelector
        };
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 2]
```diff
@@ -284,8 +284,17 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
+                    const searchEnabled = this.option('searchEnabled');
+                    const prevSelectAllEnabled = this._selectAllEnabled(previousValue);
+                    const prevItemsContainer = this._itemContainer(searchEnabled, prevSelectAllEnabled);
+
+                    this._removeClickEvent(prevItemsContainer);
+                    this._removeExpandEvent(prevItemsContainer);
+
+                    this._checkBoxModeChange(value, previousValue);
+                    this._invalidate();
+                }
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [code splitting]
It's better to leave this optionChanged handler as it is and move all your code to the `checkBoxModeChange` method.

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 3]
```diff
@@ -284,8 +284,17 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
                 }
                 break;
             case 'showCheckBoxesMode':
-                this._checkBoxModeChange(value, previousValue);
-                this._invalidate();
+                {
+                    const searchEnabled = this.option('searchEnabled');
+                    const prevSelectAllEnabled = this._selectAllEnabled(previousValue);
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [naming, minor]
Let's not use abbreviations. I understand that `prev` is a basic one, but I prefer have a common rule :)
```suggestion
                    const previousSelectAllEnabled = this._selectAllEnabled(previousValue);
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 4]
```diff
@@ -1432,6 +1442,22 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
     },
 
+    _removeClickEvent: function(itemsContainer) {
+        const { eventName, clickSelector, pointerDownEvent, pointerDownSelector } = this._getEventsData();
+
+        eventsEngine.off(itemsContainer, eventName, clickSelector);
+        eventsEngine.off(itemsContainer, pointerDownEvent, pointerDownSelector);
+    },
+
+    _getEventsData: function() {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [naming]
The name seems misleading for me - we don't get data for all events here.
```suggestion
    _getItemClickEventData: function() {
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 5]
```diff
@@ -1432,6 +1442,22 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
         });
     },
 
+    _removeClickEvent: function(itemsContainer) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [naming]
```suggestion
    _detachClickEvent: function(itemsContainer) {
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 6]
```diff
@@ -1415,15 +1428,12 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
-        const clickSelector = '.' + this._itemClass();
-        const pointerDownSelector = '.' + NODE_CLASS + ', .' + SELECT_ALL_ITEM_CLASS;
-        const eventName = addNamespace(clickEventName, this.NAME);
-        const pointerDownEvent = addNamespace(pointerEvents.down, this.NAME);
         const $itemContainer = this._itemContainer();
+        this._removeClickEvent($itemContainer);
+
+        const { eventName, clickSelector, pointerDownEvent, pointerDownSelector } = this._getEventsData();
 
         const that = this;
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [boy scout rule]
Let's get rid of `that and use arrow functions for events handlers

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 7]
```diff
@@ -830,10 +840,13 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
-        const itemSelector = this._itemSelector();
 
-        eventsEngine.off($itemsContainer, '.' + EXPAND_EVENT_NAMESPACE, itemSelector);
-        eventsEngine.on($itemsContainer, expandedEventName, itemSelector, this._expandEventHandler.bind(this));
+        this._removeExpandEvent($itemsContainer);
+        eventsEngine.on($itemsContainer, expandedEventName, this._itemSelector(), this._expandEventHandler.bind(this));
+    },
+
+    _removeExpandEvent(itemsContainer) {
+        eventsEngine.off(itemsContainer, '.' + EXPAND_EVENT_NAMESPACE, this._itemSelector());
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]:
```suggestion
        eventsEngine.off(itemsContainer, `.${EXPAND_EVENT_NAMESPACE}`, this._itemSelector());
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 8]
```diff
@@ -830,10 +840,13 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
     _initExpandEvent: function() {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: [boy scout rule]
```suggestion
    _attachExpandEvent: function() {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeViewParts/selection.js [hunk 9]
```diff
@@ -505,6 +505,34 @@ module('selection single', () => {
         treeView.checkSelected([0], items);
     });
 
+    test('selectByClick selection should work correctly after runtime "showCheckBoxesMode" option change', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/treeViewParts/selection.js]:
```suggestion
    test('selectByClick selection should work correctly after runtime "showCheckBoxesMode" option change (T1190412)', function(assert) {
```

File: packages/devextreme/js/ui/tree_view/ui.tree_view.search.js [hunk 10]
```diff
@@ -90,8 +90,8 @@ const TreeViewSearch = TreeViewBase.inherit(searchBoxMixin).inherit({
         this.$element().empty();
     },
 
-    _itemContainer: function(isSearchMode) {
-        if(this._selectAllEnabled()) {
+    _itemContainer: function(isSearchMode, selectAllEnabled) {
+        if(selectAllEnabled === undefined ? this._selectAllEnabled() : selectAllEnabled) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.search.js]: [code style]
Let's refactor a bit.
```javascript
    const selectAllEnabled ??= this._selectAllEnabled();
    if (selectAllEnabled) {
    ...
```

**Review by Zedwag — DISMISSED** [review]

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 11]
```diff
@@ -241,6 +241,13 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
     },
 
     _checkBoxModeChange: function(value, previousValue) {
+        const searchEnabled = this.option('searchEnabled');
+        const previousSelectAllEnabled = this._selectAllEnabled(previousValue);
+        const prevItemsContainer = this._itemContainer(searchEnabled, previousSelectAllEnabled);
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: missed `prevz

File: packages/devextreme/js/ui/tree_view/ui.tree_view.base.js [hunk 12]
```diff
@@ -241,6 +241,13 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
+
+        this._detachClickEvent(prevItemsContainer);
+        this._removeExpandEvent(prevItemsContainer);
```
Comment (ksercs) [inline packages/devextreme/js/ui/tree_view/ui.tree_view.base.js]: missed `remove`
```suggestion
        this._detachExpandEvent(prevItemsContainer);
```

**Review by ksercs — APPROVED** [review]

**Review by Zedwag — APPROVED** [review]

**Pull request merged**